Show/hide textbox based on answers to related question

chao-xian opened the pull request:
We only want to give the user fields to provide more detail if they answer `Yes` to a question.

If the user answers `Yes` to:

* `Should the content be removed?`, they are presented with a textbox for `Where should users be redirected to? (optional)`
* `Is this content very similar to other pages?`, they are presented with a textbox for `URLs of similar pages`

Answering `No` to either question will hide the textboxes.

## Before
![screenshot-content-performance-manager publishing service gov uk-2017-09-15-19-24-15-196](https://user-images.githubusercontent.com/424772/30497688-a81ecd88-9a4b-11e7-898f-aefe65a80c47.png)

## After 
![screenshot-localhost-3000-2017-09-15-19-24-48-478](https://user-images.githubusercontent.com/424772/30497707-bb589bf4-9a4b-11e7-9985-33763001cace.png)
![screenshot-localhost-3000-2017-09-15-19-25-09-877](https://user-images.githubusercontent.com/424772/30497711-bf4aef46-9a4b-11e7-81c9-a623342b1ae7.png)

chao-xian commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Update audit page to prevent users over-using the 'similar pages' text field](https://trello.com/c/pZJi9QwF/519-update-audit-page-to-prevent-users-over-using-the-similar-pages-text-field)

alecgibson commented on app/assets/javascripts/show-free-text-question.js:
```diff
@@ -0,0 +1,31 @@
+window.GOVUKAdmin = window.GOVUKAdmin || {};
```
I don't think we need these lines, since we've added a [file that sets these up](https://github.com/alphagov/content-performance-manager/blob/master/app/assets/javascripts/namespaces.js).

alecgibson commented on app/assets/javascripts/show-free-text-question.js:
```diff
@@ -0,0 +1,31 @@
+        var $noRadio = $element.find('.no-radio');
+        var $textBox = $element.find('.free-text-question');
+
+        if ($yesRadio.length) {
```
I don't think you need this guard?

alecgibson commented on app/assets/javascripts/show-free-text-question.js:
```diff
@@ -0,0 +1,31 @@
+
+        if ($yesRadio.length) {
+          $yesRadio.click(function(e) {
+            $textBox.removeClass('if-js-hide');
```
I personally find double-negatives hard to read - since "remove" and "hide" are both negative - it might be nice to move this into a one-line function called `showTextBox` or similar? Then you could also make this line read a bit more fluently, like:

```javascript
$yesRadio.click(showTextBox);
```

alecgibson commented on app/assets/javascripts/show-free-text-question.js:
```diff
@@ -0,0 +1,31 @@
+        }
+
+        if ($noRadio.length) {
+          $noRadio.on('click', function(e) {
```
Why have you used `click` above, but `.on('click')` here?

alecgibson commented on app/assets/javascripts/show-free-text-question.js:
```diff
@@ -0,0 +1,31 @@
+        }
+      }
+
+      addEventListener();
```
I personally prefer keeping method invocations at the top of the method, so it's clearer what's happening when you read the file.

alecgibson commented on app/assets/javascripts/show-free-text-question.js:
```diff
@@ -0,0 +1,31 @@
+  Modules.ShowFreeTextQuestion = function () {
+    this.start = function($element) {
+      function addEventListener() {
+        var $yesRadio = $element.find('.yes-radio');
```
Could you please [use `.js`-prefixed classes](https://github.com/alphagov/govuk_frontend_toolkit/blob/master/docs/javascript.md#use-js--prefixed-classes-for-interaction-hooks)?

alecgibson commented on app/views/audits/audits/_free_text_question.html.erb:
```diff
@@ -1,4 +1,4 @@
-<div class="row">
+<div class="row <% unless show_hide %>if-js-hide<% end %> t-<%= attribute.to_s.dasherize %> free-text-question">
```
`show_hide` is an oxymoron - could you please clarify what you're actually showing?

alecgibson commented on app/assets/stylesheets/_audit.scss:
```diff
@@ -24,22 +24,19 @@ $next-link-colour: #000000;
-      height: 2em;
-    }
-  }
+.t-redirect-urls,
```
What does the prefix `.t-` signify?

chao-xian replied:
I don't know - it was already a thing but as I'm not able to answer your question I've removed it.

chao-xian commented:
Thanks for your feedback @alecgibson @pmanrubia. 

@pmanrubia I've addressed your feedback about the callbacks here: https://github.com/alphagov/content-performance-manager/pull/346/commits/c008955bbf2159fd22f2de162349fbb7dda2592a

@alecgibson I addressed most of your comments in https://github.com/alphagov/content-performance-manager/pull/346/commits/6ac97e8b23e8e62ff525e6e71e0ec0efa9d6eb0c - sorry they are in all in one commit. They seemed fairly minor. Let me know if you'd prefer if I fixed them.

pmanrubia commented on spec/javascripts/fixtures/show-free-text-question.html:
```diff
@@ -0,0 +1,15 @@
+    <label for="audits_audit_redirect_urls"></label>
+    <textarea class="form-control" id="audits_audit_redirect_urls"></textarea>
+  </div>
+</div>
```
Blank line missing.

pmanrubia commented on spec/domain/audits/save_audit_spec.rb:
```diff
@@ -0,0 +1,84 @@
+    let!(:user) { create :user }
+    let!(:content_item) { create :content_item, title: "A title" }
+    let!(:filter) { Filter.new }
+    let!(:attributes) {
```
You could use `attributes_for` of `FactoryGirl`.

chao-xian replied:
Nice! Fixed in https://github.com/alphagov/content-performance-manager/pull/346/commits/55674010858f913a1056174131f50ca1a643551d#diff-dfb8ea57e9196a3e0adca3b88cec560dR5

pmanrubia commented on app/domain/audits/save_audit.rb:
```diff
@@ -0,0 +1,54 @@
+      audit.content_item = content_item
+      audit.user = user
+
+      clear_redirect_urls
```
We don't clear the fields always, so I would try to be more clear on these methods. When you read the method you could be tricked.

chao-xian replied:
Good point - thanks. Have renamed https://github.com/alphagov/content-performance-manager/pull/346/commits/bdfe179db8ee7d7f72ca0dbc282ffced02da9f34#diff-bcf8bfffc9defab9b861ec826b66da7aR43 but feels clunky.

chao-xian commented:
@pmanrubia thanks Pablo. Have addressed and fixed up.

pmanrubia commented on app/domain/audits/save_audit.rb:
```diff
@@ -0,0 +1,51 @@
+    attr_accessor :attributes, :audit, :content_item, :success, :user
+
+    def initialize(attributes:, content_id:, user_uid:)
+      self.attributes = attributes.to_h
```
to_h needs to be done outside this method.

chao-xian replied:
Thanks Pablo - addressed here: https://github.com/alphagov/content-performance-manager/pull/346/commits/a4bf9015133ca999098d1c0c99c8ec281d483f01#diff-bcf8bfffc9defab9b861ec826b66da7aR9
